### DISCUSSION TITLE
What icons are Geany using?  I would like to modify the icon theme for Linux Mint so Geany uses the Mint icons and not the fallback icons

### DISCUSSION THREAD

**Discussion by Hund** [geany/geany#3890] [General]
What icons are Geany using? I would like to modify the icon theme for Linux Mint so Geany uses the Mint icons and not the fallback icons.

Any resources available for this? I can't find anything about it.

Thanks!

**Comment by elextr** [comment]
It happens automatically here for LM Cinnamon and has since LM 20.

Edit: "It" == uses the system icon theme

**Comment by b4n** [comment]
Geany has a few specific icons that the theme might not have, like the *build* or *save all* ones.

You can find which non-standard icons Geany uses by checking out the list of icons it installs: https://github.com/geany/geany/tree/master/icons/16x16

**Comment by elextr** [comment]
@b4n, LM Cinnamon has _all_ the icons used by Geany, not a fallen over bookcase in sight ;-)

**Reply by b4n** [reply]
Including our specific ones? Neat 😊

**Reply by elextr** [reply]
The UI icons, in particular _build_ and _save-all_ which you mentioned, not the internal ones for the symbols tree of course.

**Comment by elextr** [comment]
@Hund which version of Geany and which version of LM are you using?

**Reply by Hund** [reply]
* Linux Mint 21.3 Cinnamon
* Geany 1.38. I have tried version 2.0 from Flatpak as well.

![image](https://github.com/geany/geany/assets/320498/80db04fd-f043-47fb-9f1e-4c68e683f53e)

CC: @elextr

**Reply by elextr** [reply]
Hmmm, works on the machine I am on ATM with LM 20, although Mint-Y-Sand seems to have no contrast between active and inactive icons it shows the right icons.

**Reply by b4n** [reply]
I guess the question might be "which icons aren't looking as you expect?"

**Reply by Hund** [reply]
> I guess the question might be "which icons aren't looking as you expect?"

All icons. Geany is just using some GNOME default icons.

**Reply by b4n** [reply]
That's a different question -- but it actually uses at least some of the Mint icons, as the Geany icon itself in the about dialog shows (it's the Minty one, not the regular one)

**Reply by elextr** [reply]
Also the selections of themes are mixed up, try with consistent settings, eg Mint-Y-dark for all of them and see what happens.

**Reply by Hund** [reply]
> Also the selections of themes are mixed up, try with consistent settings, eg Mint-Y-dark for all of them and see what happens.

I'm using a light theme and an icon them with dark action icons. Switching to a dark theme would result in me having issues seeing the action icons in other applications where they're actually being used. :P

**Reply by elextr** [reply]
Well, it doesn't have to be dark, make it light, and anyway its only for testing to see if it works, just make all the settings consistent.

**Reply by Hund** [reply]
Here's two images showing both the light and dark settings.

![Screenshot from 2024-06-02 12-17-54](https://github.com/geany/geany/assets/320498/05f12e38-a62e-4861-89a5-826d1de58cd7)
![Screenshot from 2024-06-02 12-18-02](https://github.com/geany/geany/assets/320498/6095d2e8-0f6b-4831-9e7b-bbc9255ea953)

**Comment by elextr** [comment]
Those are what I see if I select "Adwaita" or "hicolor" as the icon theme, sadly your images don't show the icon theme setting.

**Reply by Hund** [reply]
It shows you that I'm using the Mint-Y theme and icons (with the green version of the theme and icons). I choose the "consistent" theme settings as you told me to.

**Reply by elextr** [reply]
I can't see where the icon theme is explicitly listed in the images above, rule 1 in debugging, never assume that software is doing what you assume, check [learned from many years of angst].

If indeed the explicit icon theme is listed as "Mint-Y" it suggests that there is something wrong with your installation that is choosing the wrong icons.

**Reply by b4n** [reply]
> I can't see where the icon theme is explicitly listed in the images above

In the first image, it says "Icons: Mint-Y-Sand".

**Reply by elextr** [reply]
@b4n, yes but that image is not the same theme as the last two, which is what I was asking about.

**Reply by elextr** [reply]
LM has a lot of themes, but once I started mixing sub-themes like the first image shows I got strange inconsistent results some times, thats why I asked the OP to try a consistent theme without mixing sub-themes.

**Reply by Hund** [reply]
> LM has a lot of themes, but once I started mixing sub-themes like the first image shows I got strange inconsistent results some times, thats why I asked the OP to try a consistent theme without mixing sub-themes.

And that's what I did. I assumed you understood how the simple theme selection thingy works. :) If you choose Mint-Y like I did, it will choose the Mint-Y theme (dark or light) for everything, including the icons.

**Reply by elextr** [reply]
@Hund like I said above, when _something is going wrong_, its important to check explicitly the part you are interested in (the icon theme in this case) is what you expect.

**Reply by Hund** [reply]
> @Hund like I said above, when _something is going wrong_, its important to check explicitly the part you are interested in (the icon theme in this case) is what you expect.

I did double check it. :)

**Reply by elextr** [reply]
In that case as I said above, perhaps there is something weird in your installation.

**Reply by b4n** [reply]
Try running `GTK_DEBUG=interactive geany` and check theme and stuff in the inspector maybe?

**Reply by Hund** [reply]
I had no idea that was a thing. It looks like Mint-Y and Mint-L icon themes doesn't work for Geany. The Mint-X did change most of the icons though.

**Reply by b4n** [reply]
I can't help but ask: does it work for other apps? Geany doesn't do anything particular in this regard but for a few specific icons I mentioned, all the rest are standard names loaded by GTK, so it "should" work the same for basically all GTK apps.

**Reply by Hund** [reply]
It does work for other applications. As far as I can tell. I haven't seen the GNOME icons anywhere else, but most applications doesn't really use a toolbar any more.

**Reply by Hund** [reply]
This isn't a major issue for me though. It would have been nice to sort it out, but I don't want to spend hours debugging it either. :)

**Reply by b4n** [reply]
I had a theoretical look at this:

* [Mint-Y-Sand theme](https://github.com/linuxmint/mint-themes/blob/master/files/usr/share/themes/Mint-Y-Sand/index.theme) uses the [Mint-Y-Sand icon theme](https://github.com/linuxmint/mint-y-icons/blob/master/usr/share/icons/Mint-Y-Sand/index.theme)
* [Mint-Y-Sand icon theme](https://github.com/linuxmint/mint-y-icons/blob/master/usr/share/icons/Mint-Y-Sand/index.theme) inherits from (in order): [Mint-Y](https://github.com/linuxmint/mint-y-icons/blob/master/usr/share/icons/Mint-Y/index.theme), Adwaita, gnome, hicolor.  It provides [very few icons itself](https://github.com/linuxmint/mint-y-icons/tree/master/usr/share/icons/Mint-Y-Sand)
* [Mint-Y](https://github.com/linuxmint/mint-y-icons/blob/master/usr/share/icons/Mint-Y/index.theme) provides [more icons](https://github.com/linuxmint/mint-y-icons/tree/master/usr/share/icons/Mint-Y), but doesn't seem to provide everything we use, including e.g. neither `gtk-new` nor `document-new`.

Thus, it falls back to Adwaita for those.  So everything seems normal, except that Geany shows more icons than this theme was really meant to.

**Reply by elextr** [reply]
I finally got a chance to try on LMDE6 (my LM21 machine is still HDC).

Remember I said above I was using LM20.  In `icons/Mint-Y/actions/*` it has `geany-build`, `geany-save-all` and `geany-close-all` links to `run-build`, `document-save-all` and `window-close` respectively.  (@Hund LM20 does not have the "simplified" theme selector, just the "advanced" one, that caused some of the confusion above).

LMDE6 does not have the `geany-*` links AFAICT nor `run-build` nor `document-save-all`.

@b4n Mint-Y does provide quite a lot of icons in the `legacy` subdirectory including`document-new`, but no `geany-` ones so the fallen over bookcase and friends are back. :-(  Thats not surprising since looking at the git history it appears that most icons were removed from the `actions` subdirectory then the `legacy` one created from Adwaita `legacy`, sigh.  Presumably this is to prevent Adwaita removal of useful (legacy) stuff from breaking Mint.

I guess thats the case for Gnome anyway (except it falls back to Adwaita, so no guarantees anything legacy will remain available).

So if we had "somebody" with artistic talent and taste maybe the Geany icons could be modernised and light/dark and symbolic versions created.

But it seems to me that the whole icons thing is [bleeeep]ed up, either:
1. if an application has any action that a distro/theme doesn't have it can have no icon, or
2. the application picks the nearest available icon, like the links in LM20 above using "near enough" icons and hopes the distro/theme continues to support it, or
3. applications provide their own icons for actions that the distro doesn't support, but they will not integrate with the look and feel of many themes.

This isn't just a LM problem, and I don't have a solution, @b4n do you know what the solution is supposed to be?  Perhaps given Gnome HIG says "users must use Adwaita, they are not to be allowed to make any choices" they think thats a solution [end grumpy olde guy rant]?

**Reply by b4n** [reply]
>  Perhaps given Gnome HIG says "users must use Adwaita, they are not to be allowed to make any choices" they think thats a solution [end grumpy olde guy rant]?

AFAIK that's indeed the "solution": disallowed theming. IIUC to the extent where libadwaita (a broadly used GNOME UI library) overrides user choice.

I don't like that myself, but I don't know of an actual solution to the problem, as the app can't provide tailored support for all themes, and theme authors don't have tailored support for all apps either.

**Reply by elextr** [reply]
Hmmm, have I mentioned that its a bad idea to let desktop theme designers that know nothing about an application dictate its look and feel? Oh, I have? Ad nauseam? Oh well, add this to the list :-D

**Reply by b4n** [reply]
@elextr you're actually siding with many GNOME folks on this then, which say "don't theme our apps", because it's usually doomed to end up like this discussion where the app authors have to debug theme issues leading to frustration (and because of "branding").

I don't have a solution to this, but I like the idea people can adjust their interface to their liking, and I like consistency between apps. So I'm doomed 😁

**Comment by elextr** [comment]
Anyway after the grumpy olde folks have had their rant lets try and do something unusual and answer the OP question :grin: 

@Hund if you find Mint icons that you would like to use in place of the `geany-build` `geany-close-all` and `geany-save-all` you will need to replace the files that Mint installs of those names with `.png` and `.svg` extensions.  For LMDE6 there are files of those names in `/usr/share/icons/Tango/*/actions` and `/usr/share/icons/hicolor/*/actions`.  

Probably it would be better to check exactly what LM installs on LM21 since it might be different, I would use `synaptic`, select the `geany-common` package and `Properties` from the right click menu, then the `Installed Files` tab.  There is probably some arcane `dpkg` command line you could sudo that does the same thing.

You have to replace each file with your new one of the same resolution, ie `...16x16/actions/geanu-build.png` needs to be replaced by your new one at 16x16 resolution, then all the other resolutions and a `.svg` in the `scalable` directory.  I don't know if you could delete all the fixed resolution ones and just replace the `scalable` ones, but you probably have to regenerate the icon caches if you do.  So all in all a lot of effort, not sure its worth it.

Probably have to [update the icon caches](https://manpages.ubuntu.com/manpages/jammy/en/man1/gtk-update-icon-cache.1.html) for any change I think.

Or you could put those files in your icons theme I guess and update the icon cache for that theme.

**Reply by Hund** [reply]
Thanks! I'll look into it later on a rainy day. :)
